Implementation of feature for showing how fast server is running relative to real time, closes #9

pmacalpine opened the pull request:
…tive to real time.

This is an implementation of https://github.com/Gama11/RoboViz/issues/9.  This allows for showing the server's real time speed percentage  and is toggled with the M key when in live mode.  It's not perfect but given the information passed to the monitor I'm not sure how much better we can do.

pmacalpine commented:
I suppose that computation of the server speed might be better done in its own class outside of GameState as it's not really part of the game state.  For ease of implementation and needing access to game state values I just did everything in GameState however.

Gama11 commented:
I'm getting some ridicolously high values when no agents are connected (500-2000%, fluctuating a lot). The results are also very unstable whenever a new agent connects (reaches up to about 300%, though it then normalizes again).

pmacalpine commented:
Are you seeing this on both Windows and Linux?  Does the value change much when a game is running compared to when the game clock is stopped?  On Linux things look pretty good for me with the value quickly stabilizing to around 99-100% with no agents connected.  Also are you running in real time mode ($enableRealTimeMode set to true in rcssserver3.rb)?  With real time mode turned off I see values in the 7000% range with no agents connected.

Gama11 commented:
`$enableRealTimeMode` is true, I haven't changed the server config.

My setup is a little unusual, I'm running `rcssserver3d` in a Ubuntu Virtaul Machine and RoboViz on the Windows host machine. I doubt running RoboViz inside the VM would make a difference, but I can give it a try.

When the game is running and the values have stabilized, the values seem mostly ok.

pmacalpine commented:
When a game is not running roboviz is somewhat blindly assuming messages should arrive every 0.04 seconds (default value of $monitorStep and $renderStep in ~/.simspark/spark.rb) in real time.  If the value of 0.04 is increased roboviz will think things are running slower and if it's decreased roboviz will think the server is running faster.  Also if messages are dropped (which roboviz can't detect) there will be a longer delay between messages making roboviz think the server is running slower than it really is.  These problems shouldn't be that much of a factor when a game is running as then roboviz can look at the delta between game times to determine the speed.

It just occurred to me why you might be seeing faster times then expected: to compute the speed I'm using the average estimated speed factor between messages when I should be using a weighted average.  As an example suppose we're sending messages every second and have six messages sent at times 0, 1, 2, 3, 4, and 5 seconds.  In an ideal network these messages would arrive almost instantaneously at the same time they were sent giving us a real time speed factor between each message of 1 as well as an average speed factor of 1. In a congested network the messages might actually arrive at times 0, 3, 3.5, 4, 4.5, 5 giving us speed factors between messages of .33, 2, 2, 2, and 2 with an average speed factor of 1.67.  Both of these instances should report a speed factor of 1.  I'll look into fixing this.

pmacalpine commented:
Hopefully I have fixed this issue now where a correct average speed factor value will be calculated when there is a lot of fluctuation in time deltas between messages.  A potential issue for windows machines is that timers might not be very accurate on them https://blogs.oracle.com/dholmes/entry/inside_the_hotspot_vm_clocks.  I'm not sure whether it is better to use System.currentTimeMillis() or System.nanoTime().  Currently I'm using System.currentTimeMillis() but find both ways of measuring time to produce reasonable results when running in real time mode on linux.  When turning off real time I find System.nanoTime() to report faster times, however it looks like the server might actually be running faster then too.  I also see a bit of a delay sometimes in updates to the screen when not running in real time using System.nanoTime().  I have commented out code for using System.nanoTime() that can be turned on by searching for "_NS" and then uncommenting those lines while commenting out their "_MS" corresponding lines.  I would be curious how performance compares between using System.currentTimeMillis() and System.nanoTime() when running roboviz on windows with real time mode both turned on and off.

Gama11 commented on src/rv/comm/rcssserver/GameState.java:
```diff
@@ -324,6 +350,66 @@ public void parse(SExp exp, WorldModel world) {
+
+        Float[] deltas = serverMsgDeltas.values().toArray(new Float[0]);
+        for (int i = 1; i < deltas.length; i++) {
+            float delta = deltas[i].floatValue();
```
This call isn't necessary according to IntelliJ:

![](http://i.imgur.com/zKaVGd5.png)

Gama11 commented on src/rv/comm/rcssserver/GameState.java:
```diff
@@ -324,6 +350,66 @@ public void parse(SExp exp, WorldModel world) {
+
+        float sumDeltas = 0;
+
+        Float[] deltas = serverMsgDeltas.values().toArray(new Float[0]);
```
The `toArray()` call here is apparently not a good idea:

![](http://i.imgur.com/FDcIvsJ.png)

Gama11 commented:
Great job, the results seem very stable now with my setup. :+1: 

My observations regarding NS / MS:

- With real time mode enabled and no agents connected, the percentage is slightly higher using NS (100-102% vs 99-100%). However, this difference seems to disappear once agents are connected.
- With real time mode disabled and no agnets connnected, there is practically no noticable difference due to how high the values are and how much they fluctuate. With agents connected, the results seem pretty much identical as well.

Gama11 commented:
I'm not sure if the `Real Time` label works that well, perhaps `Server Speed` would be better?

Also, I do agree with you that this shouldn't really be in `GameState`, for several reasons - one of them being that this is only needed in live mode. Maybe we could move it into a separate `ServerSpeedBenchmarker` (feel free to come up with a better name)? `LiveGameScreen` could have an instance of that.

pmacalpine commented:
I've addressed most of the mentioned issues.  I've moved the server estimation code into its own class (added as a listener to GameState by LiveGameScreen).  I also added a static final boolean variable for switching between measuring time in milliseconds and nanoseconds.  Currently time is measured in milliseconds as I found this to be more stable than nanoseconds when not running in real time mode.

Gama11 commented:
Thanks, really happy with how this feature turned out. The implementation is now very clean as well! :)
